Extensions/RosettaTorch/Includes/AlphaZero/Trainer.hpp

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2019 Chris Ohk, Youngjoong Kim, SeungHyun Jeon
+
+// We are making my contributions/submissions to this project solely in our
+// personal capacity and are not conveying any rights to any intellectual
+// property of any third parties.
+
+// It is based on peter1591's hearthstone-ai repository.
+// References: https://github.com/peter1591/hearthstone-ai
+
+#ifndef ROSETTASTONE_TORCH_ALPHA_ZERO_TRAINER_HPP
+#define ROSETTASTONE_TORCH_ALPHA_ZERO_TRAINER_HPP
+
+#include <AlphaZero/Evaluation/Runner.hpp>
+#include <AlphaZero/Logger/ILogger.hpp>
+#include <AlphaZero/Optimizer/Runner.hpp>
+#include <AlphaZero/SelfPlay/Runner.hpp>
+#include <AlphaZero/TrainerConfig.hpp>
+#include <AlphaZero/Training/TrainingData.hpp>
+#include <AlphaZero/Utils/ThreadPool.hpp>
+#include <NeuralNet/NeuralNetwork.hpp>
+
+namespace RosettaTorch::AlphaZero
+{
+//!
+//! \brief Trainer class.
+//!
+//! This class is a trainer for target to end-user devices.
+//! Eliminate context switch as much as possible.
+//! Don't rely on threads and OS scheduling.
+//! Not targeting for distributed environments.
+//!
+class Trainer
+{
+ public:
+    //! Constructs trainer with given \p logger.
+    //! \param logger The logger to write the log.
+    Trainer(ILogger& logger);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Class 'Trainer' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/app/utilForever/RosettaStone/pullRequest?prid=4565217)